pulp_ansible/app/tasks/collectionversion_index.py

```diff
@@ -0,0 +1,193 @@
+            is_latest = True
+
+        # sometimes distros point at nothing?
+        else:
+            return
+
+    # extract repository from repository version
+    if repository is None:
+        repository = repository_version.repository
+
+    # optimization: we only want to index "distributed" CVs
+    if distribution is None and not has_distribution(repository, repository_version):
+        # clean indexes for this repository or repository version
+        CVIndex.objects.filter(
+            Q(repository=repository) | Q(repository_version=repository_version)
+        ).delete()
+        return
+
+    # We want to use the distro's repository version on the index
+    # model but we also want it to be a NoneType if the distro is
+    # using a repository instead of a repository version
+    if is_latest:
+        distribution_repository_version = None
+    else:
+        distribution_repository_version = repository_version
+
+    # what is the latest repo version?
+    repository_version_latest = repository_version
+    if not is_latest:
+        repository_version_latest = repository_version.repository.latest_version()
+
+    # optimization: exit early if the latest repository version is already indexed
+    if CVIndex.objects.filter(repository_version=repository_version_latest).exists():
+        return
+
+    # get all CVs in this repository version
+    cvs = repository_version_latest.content.filter(
+        pulp_type="ansible.collection_version"
+    ).values_list("pk", flat=True)
+    cvs = CollectionVersion.objects.filter(pk__in=cvs)
+
+    # get the set of signatures in this repo version
+    repo_signatures_pks = repository_version_latest.content.filter(
+        pulp_type="ansible.collection_signature"
+    ).values_list("pk", flat=True)
+    repo_signatures = CollectionVersionSignature.objects.filter(pk__in=repo_signatures_pks)
+
+    # get the set of deprecations in this repo version
+    deprecations = repository_version_latest.content.filter(
+        pulp_type="ansible.collection_deprecation"
+    ).values_list("pk", flat=True)
+    deprecations = AnsibleCollectionDeprecated.objects.filter(pk__in=deprecations)
+    deprecations_set = {(x.namespace, x.name) for x in deprecations}
+
+    # find all namespaces in the repo version
+    namespaces = repository_version_latest.content.filter(
+        pulp_type="ansible.namespace"
+    ).values_list("pk", flat=True)
+    namespaces = AnsibleNamespace.objects.filter(pk__in=namespaces)
+    namespaces = {x.name: x for x in namespaces}
+
+    # map out the namespace(s).name(s) for everything in the repo version
+    colset = {(x.namespace, x.name) for x in cvs}
+
+    # clean up any repo+namespace+name indexes that no longer live in the repo ...
+    for cvi in CVIndex.objects.filter(repository=repository):
+        if (cvi.collection_version.namespace, cvi.collection_version.name) not in colset:
+            cvi.delete()
+            continue
```

Comment:
```suggestion
    # map out the namespace(s).name(s) for everything in the repo version
    colset = set(cvs.values_list("namespace", "name").distinct())

    # clean up any CV indexes that no longer live in the repo ...
    CVIndex.objecst.filter(repository=repository).exclude(
        collection_version__pk__in=cvs.values_list("pk", flat=True)
    ).delete()
```
Let's let the DB do most of the work.